### PR TITLE
Feat/manifest

### LINKED ISSUE #191: gen3 version manifest
we need a way to declare versions of all our microservices and create releases for the whole product

### PR DESCRIPTION
**Pull request by frickjack**
support for applying manifests from uc-cdis/cdis-manfiest via `g3k roll deployment` - ex `g3k roll portal`

resolves #191

### CODE REVIEW

File: gen3/lib/g3k_manifest.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  if [[ (! "$1" =~ ^-*force$) && -f "${doneFilePath}" ]]; then
+    return 0
+  fi
+  if [[ ! -d "${GEN3_MANIFEST_HOME}" ]]; then
+    echo -e $(red_color "ERROR: GEN3_MANIFEST_HOME does not exist: ${GEN3_MANIFEST_HOME}")
+    echo "git clone https://github.com/uc-cdis/cdis-manifest.git ${GEN3_MANIFEST_HOME}"
+    # This will fail if proxy is not set correctly
+    git clone "https://github.com/uc-cdis/cdis-manifest.git" "${GEN3_MANIFEST_HOME}"
+  fi
+  if [[ -d "$GEN3_MANIFEST_HOME/.git" ]]; then
```
Comment (philloooo) [inline gen3/lib/g3k_manifest.sh]: how would git fetch work if `.git` doesn't exist?

File: gen3/lib/g3k_manifest.sh [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    value="${replaceMap[$key]}"
+    # this won't work if key or value contain ^ :-(
+    sed -i.bak "s^${key}^${value}^g" "$tempFile"
+  done
+  cat $tempFile
+  /bin/rm "$tempFile"
+  return 0
+}
+
+g3k_roll() {
```
Comment (philloooo) [inline gen3/lib/g3k_manifest.sh]: lots of our features include multiple services upgrade, can we support a `roll all`?

File: gen3/lib/g3k_manifest.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+GEN3_HOME="${GEN3_HOME:-$(cd "${G3K_MANIFEST_SH_DIR}/../.." && pwd)}"
+GEN3_MANIFEST_HOME="${GEN3_MANIFEST_HOME:-"$(cd "${GEN3_HOME}/.." && pwd)/cdis-manifest"}"
+export GEN3_HOME
+export GEN3_MANIFEST_HOME
+
+#
+# Internal init helper
+#
+g3k_manifest_init() {
+  # do this at most once a day
```
Comment (philloooo) [inline gen3/lib/g3k_manifest.sh]: are you gonna use it for jenkins? do you want to support parameter to supply branch to use?

File: gen3/lib/shunit.sh [hunk 4]
```diff
@@ -62,7 +62,7 @@ shunit_runtest() {
   echo -e "\nRunning $testName"
   SHUNIT_CURRENT_TEST="$testName"
   let SHUNIT_TEST_COUNT+=1
-  let result=0
+  let result=0 || true
```
Comment (philloooo) [inline gen3/lib/shunit.sh]: what's this for?

File: gen3/lib/testData/default/expectedFenceResult.yaml [hunk 5]
```diff
@@ -0,0 +1,81 @@
+apiVersion: extensions/v1beta1
```
Comment (philloooo) [inline gen3/lib/testData/default/expectedFenceResult.yaml]: what are these test yamls for?

Comment (frickjack) [reply gen3/lib/testData/default/expectedFenceResult.yaml]: `g3k testsuite` uses those - the Jenkinsfile runs that test suite ... gen3/lib/g3k_testsuite.sh ...
the kube.sh/g3k/gen3 scripts are starting to share some logic

File: gen3/lib/shunit.sh (hunk 4 above)
Comment (frickjack) [reply gen3/lib/shunit.sh]: Avoid exit code 1:
   https://unix.stackexchange.com/questions/32250/why-does-a-0-let-a-return-exit-code-1

File: gen3/lib/g3k_manifest.sh (hunk 3 above)
Comment (frickjack) [reply gen3/lib/g3k_manifest.sh]: That's just there to automatically checkout/update cdis-manifest/ if it's not already done.  Jenkins will ensure that it runs against the latest manifest independent of this `g3k_manifest_init` - it's just there so things work for a basic user ...

File: gen3/lib/g3k_manifest.sh (hunk 1 above)
Comment (frickjack) [reply gen3/lib/g3k_manifest.sh]: yeah - if the folder is not managed by `git`, then don't try to do a `git fetch`.
this guard only comes into play when we're running the test suite - where we point the code at a folder with some staged test manifests ...

File: gen3/lib/g3k_manifest.sh (hunk 2 above)
Comment (frickjack) [reply gen3/lib/g3k_manifest.sh]: ok - I'll give it a try :-)

File: gen3/lib/g3k_manifest.sh (hunk 1 above)
Comment (philloooo) [reply gen3/lib/g3k_manifest.sh]: oh sorry, read it in the other way

File: gen3/lib/shunit.sh (hunk 4 above)
Comment (philloooo) [reply gen3/lib/shunit.sh]: o.o good to know, thanks

**Pull request merged**